CMPUT404W22AMNRY/CMPUT404-project-socialdistribution
# Upload Images to Amazon AWS S3

### Overview
- Add functionality to return encoded images that are links to images from the `/image` API endpoint
- Change image URL validator to accept any `Content-Type: image/*` instead of guessing by file extension
    - This will allow for links to images such as GitHub profile pictures 
- Setup `cmput404-heroku`, `cmput404-dev` and `cmput404-test` Amazon AWS S3 buckets
- Setup support for `.env` to hold configuration files for dev and heroku
- Add environment variables to Github Actions
- Add requirements for S3
- Add requirements for `.env`

### How to Test
- Add `.env` file to the project root of your directory containing the details for AWS authentication/configuration (we can probably change this to use the GitHub keyvault later) 
- `pip install -r requirements.txt`
- `python manage.py migrate`
- If there are any existing photos in local `posts_post` database and you want to retain them, you can upload them to the `cmput404-dev` bucket in `/media/images`, and then run this SQL command to keep them compatible with the new storage system:
```SQL
    UPDATE posts_post SET img_content = REPLACE(img_content, 'images/', '');
```
- Otherwise, upload a new image the same as before, right click and select `open image in new tab` to see its being hosted on AWS. For example, [here](https://cmput404-dev.s3.amazonaws.com/media/images/a31dc2c11e944ca0971d6abe711b8dfd.jpg?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA4HM362I6H3HHJ7AO%2F20220324%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20220324T214036Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=c059efb330bf84784a684a4751dc96eefd88b4c2f4711136f2444c24897dcae8) is a link to an uploaded image file that is publicly accessible. 

### Setup for Heroku
- [Configure credentials](https://devcenter.heroku.com/articles/s3#setting-up-s3-for-your-heroku-app)

Closes #108
